XMonad/Util/ClickableWorkspaces.hs

```diff
@@ -18,44 +18,72 @@ module XMonad.Util.ClickableWorkspaces (
   -- * Usage
   -- $usage
   clickablePP,
-  clickableWrap
+  clickableRenamedPP,
+  clickableWrap,
+
+  -- * Integrations
+  clickableWorkspaceNamesPP,
+  clickableMarshallPP,
   ) where
 
+import Data.Functor ((<&>))
+
 import XMonad
-import XMonad.Util.WorkspaceCompare (getWsIndex)
+import XMonad.Actions.WorkspaceNames
 import XMonad.Hooks.DynamicLog (xmobarAction, xmobarRaw, PP(..))
+import XMonad.Layout.IndependentScreens
+import XMonad.Util.WorkspaceCompare (getWsIndex)
 
 -- $usage
 -- However you have set up your PP, apply @clickablePP@ to it, and bind the result
 -- to "XMonad.Hooks.DynamicLog"\'s dynamicLogWithPP like so:
--- 
+--
 -- > logHook = clickablePP xmobarPP { ... } >>= dynamicLogWithPP
 --
 -- * Requirements:
---   * wmctrl on system (in path)
---   * "XMonad.Hooks.EwmhDesktops" for wmctrl support (see Hackage docs for setup)
---   * use of UnsafeStdinReader in xmobarrc (rather than StdinReader)
+--   * @xdotool@ on system (in path)
+--   * "XMonad.Hooks.EwmhDesktops" for @xdotool@ support (see Hackage docs for setup)
+--   * use of UnsafeStdinReader/UnsafeXMonadLog in xmobarrc (rather than StdinReader/XMonadLog)
 
 
 clickableWrap :: Int -> String -> String
-clickableWrap i ws = xmobarAction ("wmctrl -s " ++ show i) "1" $ xmobarRaw ws
+clickableWrap i ws = xmobarAction ("xdotool set_desktop " ++ show i) "1" $ xmobarRaw ws
 
--- Use index of workspace in users config to target workspace with wmctrl switch.
-getClickable :: X (WorkspaceId -> String)
-getClickable = do
+-- | Use index of workspace in users config to target workspace with @xdotool@ switch.
+getClickable :: (WorkspaceId -> String) -> X (WorkspaceId -> String)
+getClickable ren = do
   wsIndex <- getWsIndex
   return $ \ws -> case wsIndex ws of
```

Comment:
In order to get IndepenentScreens and WorkspaceNames to play nicely together, I had to apply the rename at a different point,
```haskell
return $ \ws -> case wsIndex ws of
    case wsIndex (ren ws)
        Just idx -> clickableWrap idx ws
        Nothing -> ws
```
Maybe that's related to https://github.com/xmonad/xmonad-contrib/pull/462/files#r569591764

Reply:
Ah, in my case the more appropriate type for `ren` was `String -> WorkspaceId`.

Reply:
Yeah I remember we talked about this in https://github.com/xmonad/xmonad-contrib/pull/390 and I was confused then, but I think I understand it better now. You had to apply `ren` (`marshall`) before `wsIndex` because the `marshallPP` already mangled the workspace names for you. But `clickableMarshallPP` uses `unmarshall` only after the workspace name is looked up, so it should work well.

The only issue here, I believe, is that you use both IndependentWorkspaces and WorkspaceNames and there simply isn't any `clickableMarshallWorkspaceNamesPP`. :-)

I'll add one.

Reply:
I pushed a `clickableMarshallWorkspaceNamesPP` which should make almost all of your https://github.com/ivanbrennan/nixbox/blob/c10c2d61b8d1bfffa0fb05afe721a6cf88e91844/services/xserver/xmonad.hs#L382-L393 and https://github.com/ivanbrennan/nixbox/blob/c10c2d61b8d1bfffa0fb05afe721a6cf88e91844/services/xserver/xmonad.hs#L418-L429 unnecessary. Just stick it in `logString`, compose with `namedScratchpadFilterOutWorkspacePP`, and I hope it'll work.

Reply:
@liskin Thanks! I tried it out and it all seems to work perfectly now.

Reply:
Awesome!